src/histogram.rs

```diff
@@ -14,147 +15,169 @@ const DSER_ERROR: &str = "Couldn't deserialize object";
-                let dbfile = Path::new(db).join(Path::new("store"));
-                if dbfile.exists() {
+                let dbfile = db.join("store");
+                let hm = if dbfile.exists() {
                     debug!("Zenith store exists, opening...");
-                    load_zenith_store(dbfile, &current_time)
+                    load_zenith_store(&dbfile, &current_time)
                 } else {
+                    None
+                }
+                .unwrap_or_else(|| {
+                    if let Err(e) = fs::remove_file(dbfile) {
+                        error!("{}", e);
+                    }
                     debug!("Starting a new database.");
                     HistogramMap {
                         map: HashMap::with_capacity(5),
                         duration: dur,
                         tick,
-                        db: path,
+                        db: Some(db),
                         previous_stop: None,
                     }
-                }
+                });
+                hm
             }
             None => {
                 debug!("Starting with no DB.");
                 HistogramMap {
                     map: HashMap::with_capacity(5),
                     duration: dur,
                     tick,
-                    db: path,
+                    db: None,
                     previous_stop: None,
                 }
             }
         }
     }
 
-    pub fn get_zoomed(
-        &self,
-        name: &str,
-        zoom_factor: u32,
-        update_number: u32,
-        width: usize,
-        offset: usize,
-    ) -> Option<Histogram> {
+    pub fn get_zoomed<'a>(&'a self, name: &HistogramKind, view: &View) -> Option<Histogram<'a>> {
         let h = self.get(name)?;
-
-        let mut nh = Histogram::new(width);
-        let mut h = h.clone();
-        for _i in 0..zoom_factor as usize * offset {
-            h.data.pop();
+        let h_data = h.data();
+        let h_len = h_data.len();
+
+        if view.zoom_factor == 1 {
+            let low = h_len - (view.width + view.offset);
+            let high = h_len - view.offset;
+            return Some(Histogram {
+                data: Cow::Borrowed(&h_data[low..high]),
+            });
         }
-        let nh_len = nh.data.len();
-        let zf = zoom_factor as usize;
-        let mut si: usize = if (width * zf) > h.data.len() {
-            0
-        } else {
-            h.data.len() - (width * zf) - update_number as usize
-        };
 
-        for index in 0..nh_len {
-            if si + zf <= h.data.len() {
-                nh.data[index] = h.data[si..si + zf].iter().sum::<u64>();
-            } else {
-                nh.data[index] = h.data[si..].iter().sum::<u64>();
-            }
-            si += zf;
-        }
+        let zf = view.zoom_factor as usize;
+        let start: usize =
+            h_len.saturating_sub((view.width + view.offset) * zf + view.update_number as usize);
```

Comment:
Ohh, always great when I learn something new! Using a saturating sub here is great.